docs/ABI.rst

```diff
@@ -1005,181 +1060,187 @@ mangled in to disambiguate.
-the outermost depth::
-
-  urFq_q_                           // <T_0_0> T_0_0 -> T_0_0
-  u_0_rFq_qd_0_                     // <T_0_0><T_1_0, T_1_1> T_0_0 -> T_1_1
-
-Value Witnesses
-~~~~~~~~~~~~~~~
+  generic-signature ::= requirement* 'r' GENERIC-PARAM-COUNT*
 
-TODO: document these
+  GENERIC-PARAM-COUNT ::= 'z'                // zero parameters
+  GENERIC-PARAM-COUNT ::= INDEX              // N+1 parameters
 
-::
+  requirement ::= req-spec
+  requirement ::= assoc-type req-spec        // use GENERIC-PARAM-INDEX of assoc-type
 
-  value-witness-kind ::= 'al'           // allocateBuffer
-  value-witness-kind ::= 'ca'           // assignWithCopy
-  value-witness-kind ::= 'ta'           // assignWithTake
-  value-witness-kind ::= 'de'           // deallocateBuffer
-  value-witness-kind ::= 'xx'           // destroy
-  value-witness-kind ::= 'XX'           // destroyBuffer
-  value-witness-kind ::= 'Xx'           // destroyArray
-  value-witness-kind ::= 'CP'           // initializeBufferWithCopyOfBuffer
-  value-witness-kind ::= 'Cp'           // initializeBufferWithCopy
-  value-witness-kind ::= 'cp'           // initializeWithCopy
-  value-witness-kind ::= 'TK'           // initializeBufferWithTakeOfBuffer
-  value-witness-kind ::= 'Tk'           // initializeBufferWithTake
-  value-witness-kind ::= 'tk'           // initializeWithTake
-  value-witness-kind ::= 'pr'           // projectBuffer
-  value-witness-kind ::= 'xs'           // storeExtraInhabitant
-  value-witness-kind ::= 'xg'           // getExtraInhabitantIndex
-  value-witness-kind ::= 'Cc'           // initializeArrayWithCopy
-  value-witness-kind ::= 'Tt'           // initializeArrayWithTakeFrontToBack
-  value-witness-kind ::= 'tT'           // initializeArrayWithTakeBackToFront
-  value-witness-kind ::= 'ug'           // getEnumTag
-  value-witness-kind ::= 'up'           // destructiveProjectEnumData
-  value-witness-kind ::= 'ui'           // destructiveInjectEnumTag
-
-``<value-witness-kind>`` differentiates the kinds of value
-witness functions for a type.
+  req-spec ::= protocol 'Rp' GENERIC-PARAM-INDEX // protocol requirement
+  req-spec ::= type 'Rb' GENERIC-PARAM-INDEX     // base class requirement, type is a nominal class or a substitution
+  req-spec ::= type 'Rs' GENERIC-PARAM-INDEX     // same-type requirement
+
+  GENERIC-PARAM-INDEX ::= '_'                // depth = 0,   idx = 0
+  GENERIC-PARAM-INDEX ::= INDEX              // depth = 0,   idx = N+1
+  GENERIC-PARAM-INDEX ::= 'd' INDEX INDEX    // depth = M+1, idx = N
+
+
+A generic signature begins with an optional list of requirements.
+The ``<GENERIC-PARAM-COUNT>`` describes the number of generic parameters at
+each depth of the signature. As a special case, no ``<GENERIC-PARAM-COUNT>``
+values indicates a single generic parameter at the outermost depth::
+
+  q_q_cru                           // <T_0_0> T_0_0 -> T_0_0
+  q_qd_0_cr_0_u                     // <T_0_0><T_1_0, T_1_1> T_0_0 -> T_1_1
```

Comment:
Having a single character encoding for `t_0_0` (`x` in the old scheme) was a pretty significant size win, since a large number of generics have a single generic parameter. Is there room to do that in this scheme?